packages/cli/src/credentials/credentials.service.ts

```diff
@@ -1,31 +1,91 @@
+/* eslint-disable no-restricted-syntax */
 /* eslint-disable import/no-cycle */
-import { Credentials } from 'n8n-core';
-import { FindOneOptions } from 'typeorm';
+import { Credentials, UserSettings } from 'n8n-core';
+import {
+	ICredentialDataDecryptedObject,
+	ICredentialsDecrypted,
+	INodeCredentialTestResult,
+	LoggerProxy,
+} from 'n8n-workflow';
+import { FindOneOptions, In } from 'typeorm';
+import { clone } from 'lodash';
 
-import { Db } from '..';
+import {
+	createCredentialsFromCredentialsEntity,
+	CredentialsHelper,
+	Db,
+	ICredentialsDb,
+	ResponseHelper,
+	whereClause,
+} from '..';
+import { RESPONSE_ERROR_MESSAGES } from '../constants';
 import { CredentialsEntity } from '../databases/entities/CredentialsEntity';
 import { SharedCredentials } from '../databases/entities/SharedCredentials';
+import { User } from '../databases/entities/User';
+import { validateEntity } from '../GenericHelpers';
+import { CredentialRequest } from '../requests';
+import { externalHooks } from '../Server';
 
 export class CredentialsService {
-	static async getSharing(
-		userId: string,
+	static async getSharedCredentials(
+		user: User,
 		credentialId: number | string,
 		relations?: string[],
+		allowGlobalOwner = true,
 	): Promise<SharedCredentials | undefined> {
 		const options: FindOneOptions = {
 			where: {
-				user: { id: userId },
 				credentials: { id: credentialId },
 			},
 		};
 
+		if (!allowGlobalOwner || user.globalRole.name !== 'owner') {
+			// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+			options.where.user = { id: user.id };
+		}
+
```

Comment:
If `relations` is an empty array, the join still happens? If so, we should check for length.